core/src/storage/collections/btree_map/impls.rs

```diff
@@ -0,0 +1,1963 @@
+/// (a node with children). This distinction makes it easier to handle e.g.
+/// cases where nodes are merged or implementing recursion.
+pub(super) enum HandleType {
+    Leaf,
+    Internal,
+}
+
+/// This enum is used when recursively processing an underfull node (i.e. a node has so
+/// few elements that it could be merged with another one). This happens when a key/value
+/// pair is removed from a node.
+///
+/// This enum is returned as a result of `.handle_underfull_node`, which contains
+/// the underfull handling logic. Based on this result the recursion either proceeds or stops.
+enum UnderflowResult {
+    /// We are at the root now.
+    AtRoot,
+    /// The node has an empty parent, referred to by the `NodeHandle` in this tuple.
+    EmptyParent(NodeHandle),
+    /// The elements of an underfull node were merged into another node. The contained
+    /// `NodeHandle` points to a parent node. The right node was merged into the left
+    /// node and subsequently removed.
+    Merged(NodeHandle),
+    /// If merging is not possible we "steal" an element from one node and put it into
+    /// its neighboring node.
+    Stole(NodeHandle),
+}
+
+/// Mapping stored in the contract storage.
+///
+/// This implementation follows the algorithm used by the Rust
+/// `BTreeMap` standard library implementation. The Rust implementation
+/// is in-memory, whereas this implementation uses the ink! storage
+/// primitives (`SyncChunk`, etc.).
+#[cfg_attr(feature = "ink-generate-abi", derive(Metadata))]
+pub struct BTreeMap<K, V> {
+    /// Stores densely packed general BTreeMap information.
+    header: storage::Value<BTreeMapHeader>,
+
+    /// Contains the nodes of the tree.
+    entries: SyncChunk<InternalEntry<K, V>>,
+
+    /// The key/value pairs stored in the tree.
+    /// We don't store this in `entries` for performance reasons. Instead each
+    /// node of the tree merely stores a pointer to the storage index where
+    /// the key/value pair is store.
+    /// This way it's more effective to balance the tree, since we only ever
+    /// have to move the `u32` index, instead of the whole `K`/`V` (which
+    /// would imply decoding, encoding, etc.).
+    kv_pairs: SyncChunk<InternalKVEntry<K, V>>,
+}
+
+/// Densely stored general information required by a map.
+///
+/// # Note
+///
+/// Separation of these fields into a sub structure has been made
+/// for performance reasons so that they all reside in the same
+/// storage entity. This allows implementations to perform less reads
+/// and writes to the underlying contract storage.
+#[derive(Encode, Decode)]
+#[cfg_attr(feature = "ink-generate-abi", derive(Metadata))]
+struct BTreeMapHeader {
+    /// The latest vacant node index.
+    next_vacant: Option<NodeHandle>,
+    /// The latest vacant pair index.
+    next_vacant_pair: Option<KVStoragePointer>,
+    /// The index of the root node.
+    root: Option<NodeHandle>,
```

Comment:
Is this not always 0? If not, when? why?

Reply:
The root index can change. For example, if an element is inserted it is inserted into a leaf; if the leaf is already full this results in changes to the tree structure. These changes are propagated upwards. This might result in an overfull root node. The algorithm handles this by creating a new root node with the old root as a child.